dbms/src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -43,51 +71,162 @@ struct GroupArrayNumericData
     Array value;
 };
 
+template <typename T>
+struct GroupArrayNumericData<T, true>
+{
+    // Switch to ordinary Allocator after 4096 bytes to avoid fragmentation and trash in Arena
+    using Allocator = MixedAlignedArenaAllocator<alignof(T), 4096>;
+    using Array = PODArray<T, 32, Allocator>;
+
+    Array value;
+    size_t total_values = 0;
+    pcg32_fast rng;
+
+    UInt64 genRandom(size_t lim)
+    {
+        /// With a large number of values, we will generate random numbers several times slower.
+        if (lim <= static_cast<UInt64>(rng.max()))
+            return static_cast<UInt32>(rng()) % static_cast<UInt32>(lim);
+        else
+            return (static_cast<UInt64>(rng()) * (static_cast<UInt64>(rng.max()) + 1ULL) + static_cast<UInt64>(rng())) % lim;
```

Comment:
What about just using pcg64?

Reply:
It's borrowed from https://github.com/yandex/ClickHouse/blob/master/dbms/src/AggregateFunctions/ReservoirSampler.h#L205  and I noticed a `fast` suffix... I thought this was an optimization you did in be79166199fc0ee9c04a622f3894f2d52ee14448

Reply:
be79166 was just a modernization of code base. Without much focus on performance on this particular function.